contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/utils/structs/DoubleEndedQueueUpgradeable.sol";
+
+import "./StBNB.sol";
+import "./SystemConfig.sol";
+
+interface IStakeHub {
+    function isPaused() external view returns (bool);
+    function getUnbondTime() external view returns (uint256);
+}
+
+contract StakePool is SystemConfig, StBNB {
+    using DoubleEndedQueueUpgradeable for DoubleEndedQueueUpgradeable.Bytes32Deque;
+
+    /*----------------- storage -----------------*/
+    address public validator;
+
+    uint256 private _totalReceivedReward; // just for statistics
+    uint256 private _totalPooledBNB; // total reward plus total BNB staked in the pool
+
+    // hash of the unbond request => unbond request
+    mapping(bytes32 => UnbondRequest) private _unbondRequests;
+    // user => unbond request queue(hash of the request)
+    mapping(address => DoubleEndedQueueUpgradeable.Bytes32Deque) private _unbondRequestsQueue;
+    // user => locked shares
+    mapping(address => uint256) private _lockedShares;
+
+    struct UnbondRequest {
+        uint256 amount;
+        uint256 unlockTime;
+    }
+
+    /*----------------- events -----------------*/
+    event Delegated(address indexed sender, uint256 amount);
+    event RewardReceived(uint256 amount);
+    event UnbondRequested(address indexed sender, uint256 amount, uint256 unlockTime);
+    event UnbondClaimed(address indexed sender, uint256 amount);
+
+    /*----------------- modifiers -----------------*/
+    modifier onlyStakeHub() {
+        address sender = _msgSender();
+        require(sender == STAKE_HUB, "NOT_STAKE_HUB");
+        _;
+    }
+
+    modifier onlyValidatorSet() {
+        address sender = _msgSender();
+        require(sender == VALIDATOR_SET, "NOT_VALIDATOR_SET");
+        _;
+    }
+
+    modifier whenNotPaused() {
+        require(!IStakeHub(STAKE_HUB).isPaused(), "CONTRACT_IS_STOPPED");
+        _;
+    }
+
+    /*----------------- external functions -----------------*/
+    function initialize(address _validator, uint256 _selfDelegateAmt) public payable initializer {
+        validator = _validator;
+
+        _bootstrapInitialHolder(_selfDelegateAmt);
+    }
+
+    function delegate(
+        address _delegator,
+        uint256 _bnbAmount
+    ) external payable onlyStakeHub whenNotPaused returns (uint256) {
+        return _stake(_delegator, _bnbAmount);
+    }
+
+    function undelegate(address _delegator, uint256 _sharesAmount) external onlyStakeHub whenNotPaused {
+        require(_sharesAmount != 0, "ZERO_UNDELEGATE");
+        require(_sharesAmount <= _sharesOf(_delegator), "INSUFFICIENT_BALANCE");
+
+        // lock the tokens
+        _transfer(_delegator, address(this), _sharesAmount);
```

Comment:
why transfer to `this` rather than burn directly?

Reply:
what is the bad if change the `totalShares `?

Reply:
Done